crates/shared/src/price_estimation/trade_verifier/balance_overrides/detector.rs

```diff
@@ -39,121 +43,287 @@ impl std::ops::Deref for Detector {
-        let state_overrides = strategies
-            .iter()
-            .map(|helper| helper.strategy.state_override(&holder, &helper.balance))
-            .collect::<HashMap<_, _>>();
+    // Reverse so most recently accessed come first
+    strategies.reverse();
+
+    // Extract just the strategies
+    strategies
+        .into_iter()
+        .map(|(strategy, _)| strategy)
+        .collect()
+}
 
+impl Detector {
+    /// Creates a new balance override detector.
+    pub fn new(web3: ethrpc::Web3, probing_depth: u8) -> Self {
         Self(Arc::new(Inner {
             web3,
-            holder,
-            strategies,
-            state_overrides,
+            probing_depth,
         }))
     }
 
-    /// Tries to detect the balance override strategy for the specified token.
-    /// Returns an `Err` if it cannot detect the strategy or an internal
-    /// simulation fails.
-    pub async fn detect(&self, token: Address) -> Result<Strategy, DetectionError> {
-        let token = ERC20::Instance::new(token.into_alloy(), self.web3.alloy.clone());
-        let overrides = hashmap! {
-            token.address().into_legacy() => StateOverride {
-                state_diff: Some(self.state_overrides.clone()),
-                ..Default::default()
-            },
+    /// Detects the balance storage slot using debug_traceCall, similar to
+    /// Foundry's `deal`. This traces a balanceOf call and finds which
+    /// storage slot is accessed. If more than one slot is accessed, we reverse
+    /// by order accessed, sort by if the slot was one that would normally
+    /// be seen in solidity mapping, and test one by one to see if the
+    /// override is effective.
+    pub async fn detect(
+        &self,
+        token: Address,
+        holder: Address,
+    ) -> Result<Strategy, DetectionError> {
+        let balance_of_call = ERC20::ERC20::balanceOfCall { account: holder };
+        let calldata = balance_of_call.abi_encode();
+
+        let call_request = TransactionRequest {
+            to: Some(TxKind::Call(token)),
+            input: TransactionInput::new(calldata.into()),
+            ..Default::default()
         };
-        let balance = token
-            .balanceOf(self.holder.into_alloy())
-            .state(overrides.into_alloy())
-            .call()
+
+        let trace = self
+            .web3
+            .alloy
+            .debug_trace_call(
+                call_request,
+                BlockId::latest(),
+                GethDebugTracingCallOptions::default(),
+            )
             .await
-            .context("eth_call with state overrides failed")
-            .map_err(|e| DetectionError::Simulation(SimulationError::Other(e)))?;
+            .map_err(|e| {
+                tracing::debug!(?token, error = ?e, "debug_traceCall not supported for token");
+                DetectionError::Simulation(SimulationError::Other(anyhow::anyhow!(
+                    "debug_traceCall failed: {e}"
+                )))
```

Comment:
`err` is used across the codebase. Also, this can probably be improved if the underlying error can be easily converted into alloy.
```suggestion
            .map_err(|err| {
                tracing::debug!(?token, ?err, "debug_traceCall not supported for token");
                DetectionError::Simulation(SimulationError::Other(err.into()))
                .context("debug_traceCall failed")
```
Also, do we really need to log here? Will the error be logged somewhere on the caller side?

Reply:
I wanted to add a log here because if we are trying to debug why we cant do a balance override for a certain token, having a log message emitting from the source of the problem (even if it bubbles into a full blown error somewhere else later) is super useful.

Reply:
ok so I ran into an issue getting he error like you suggested since the type that we use for detection error is `thiserror`, so what I did is added a new `DetectionError::Rpc` type which will handle any Rpc errors that come from alloy (seemed like a new enough type of error to warrent a growth of the enum) bd15a3ac4